Ports/RISC-V/RV32/GCC/os_cpu_a.S

```diff
@@ -153,19 +170,94 @@ OSStartHighRdy:
 #              the result of an interrupt.  This function triggers a synchronous software interrupt by
 #              writing into the MSIP register
 #********************************************************************************************************
+OSIntCtxSw:
+    ret
 
 OSCtxSw:
-OSIntCtxSw:
-# MIE_MSIE -- enable software interrupt bit
-    li     t0, RISCV_MIE_MSIE
-    csrrs  zero, mie, t0
-
-# This will trigger a synchronous software interrupt; PRCI->MSIP[0] = 0x01;
-    li     t0, RISCV_PRCI_BASE_ADDR
-    li     t1, 0x1
-    sw     t1, 0x0(t0)
+    ecall
     ret
 
+# Exception handler should be 8 byte aligned
+.align 8
+ucos_intr_exception_handler:
+    # save regs to current sp
+    addi sp, sp, -4*32
+    # store contents of register ra into memory [(0 * 4) + sp]
+    sw     ra,   0 * 4(sp)
+    sw     t0,   4 * 4(sp)
+    sw     t1,   5 * 4(sp)
+    sw     t2,   6 * 4(sp)
+    sw     s0,   7 * 4(sp)
+    sw     s1,   8 * 4(sp)
+    sw     a0,   9 * 4(sp)
+    sw     a1,  10 * 4(sp)
+    sw     a2,  11 * 4(sp)
+    sw     a3,  12 * 4(sp)
+    sw     a4,  13 * 4(sp)
+    sw     a5,  14 * 4(sp)
+    sw     a6,  15 * 4(sp)
+    sw     a7,  16 * 4(sp)
+    sw     s2,  17 * 4(sp)
+    sw     s3,  18 * 4(sp)
+    sw     s4,  19 * 4(sp)
+    sw     s5,  20 * 4(sp)
+    sw     s6,  21 * 4(sp)
+    sw     s7,  22 * 4(sp)
+    sw     s8,  23 * 4(sp)
+    sw     s9,  24 * 4(sp)
+    sw     s10, 25 * 4(sp)
+    sw     s11, 26 * 4(sp)
+    sw     t3,  27 * 4(sp)
+    sw     t4,  28 * 4(sp)
+    sw     t5,  29 * 4(sp)
+    sw     t6,  30 * 4(sp)
+
+    # If it is a interrupt, do not add 4 in mepc
+    # If it is a exception, add 4 in mepc
+    li t1, ARCH_CPU_MCAUSE_CAUSE_MASK
+    csrr  t0, mcause
+    # t3 = t1 & t0
+    and t3, t1, t0
+    li t1, 11
+    csrr   t0,  mepc
+    # If t1 != t3 jump to DONTADD:
+    bne t1, t3, DONTADD
+    addi   t0,  t0, 4
+   
+DONTADD: # for interrupt we dont need to add 4 in sp
+    sw     t0,  31*4(sp)
+    # load address of OSTCBCurPtr into t0 register
+    la     t0, OSTCBCurPtr
+    # load word from register [t0 + 0] address into t1
+    lw     t1, 0(t0)
+    # store value stored in sp to [t1 + 0] address
+    sw     sp, 0(t1)
+
+    li t1, ARCH_CPU_MCAUSE_INTR_MASK
+    csrr  t0, mcause
+    and t2, t1, t0
+    li t1, 1
```

Comment:
instead of 1, shouldnt this be ARCH_CPU_MCAUSE_INTR_MASK

Reply:
And t1 is already loaded on top with ARCH_CPU_MCAUSE_INTR_MASK

 so maybe this line is not needed